### ISSUE TITLE
Data leakage during validation

### ISSUE THREAD

**Issue by bradand** [hachanook/pyinn#3]
In the training code for the Regression_INN class and its subclasses, there is an unintended data leakage during the validation and testing phases.

During training, the model parameters are correctly updated using the training data via the update_optax method, which computes the loss, gradients, and applies updates. However, in the validation loop (which runs every validation_period epochs) and the final testing phase, the code also calls update_optax on the validation and test batches, respectively. This means the model is inadvertently computing gradients from the validation/test losses and updating its parameters/opt_state based on those datasets. As a result, validation and test data "leak" into the training process, which can lead to overly optimistic performance metrics, unreliable model evaluation, and potential overfitting to the holdout sets. This violates the principle of keeping validation and test data isolated from training to ensure fair assessment of generalization.
The issue appears in these lines:

Validation: _, _, loss_val, u_pred_val = self.update_optax(params, opt_state, x_val, u_val)
Testing: _, _, _, u_pred_test = self.update_optax(params, opt_state, x_test, u_test)

This affects all derived classes (e.g., Regression_INN_sequential, Regression_MLP, Classification_INN, Classification_MLP) since they inherit or override the train method similarly.

Suggested Fix
To resolve this, replace the calls to update_optax in the validation and testing loops with direct calls to get_loss, which computes the loss and predictions without computing gradients or updating parameters. This keeps the evaluation phases purely inference-based.

Here's a patch-like suggestion for the train method in Regression_INN (apply similarly to subclasses if overridden):

In the validation block (inside if (epoch+1) % self.validation_period == 0):
```

error_cum, ndata_cum = 0, 0
for batch in self.cls_data.val_dataloader:
    x_val, u_val = jnp.array(batch[0]), jnp.array(batch[1])
    loss_val, u_pred_val = self.get_loss(params, x_val, u_val)  # <-- Change here: use get_loss instead of update_optax
    error_cum, ndata_cum = self.get_error_metrics(u_val, u_pred_val, error_cum, ndata_cum)

```
In the test block (after training loop):

```
error_cum, ndata_cum = 0, 0
for batch in self.cls_data.test_dataloader:
    x_test, u_test = jnp.array(batch[0]), jnp.array(batch[1])
    _, u_pred_test = self.get_loss(params, x_test, u_test)  # <-- Change here: use get_loss and ignore loss if not needed
    error_cum, ndata_cum = self.get_error_metrics(u_test, u_pred_test, error_cum, ndata_cum)

```
Notes:

get_loss returns (loss, u_pred), so you can unpack accordingly (and ignore loss in testing if it's not used).
No changes are needed to opt_state in these phases, as it's irrelevant for evaluation.

This fix should also be verified in the inference method, where multiple redundant calls to forward exist—consider consolidating to a single call for efficiency, but it's not directly related to the leakage.
